src/hugr/views/sibling.rs

```diff
@@ -0,0 +1,575 @@
+//! Views for HUGR sibling subgraphs.
+//!
+//! Views into subgraphs of HUGRs within a single level of the
+//! hierarchy, i.e. within a sibling graph. Such a subgraph is
+//! represented by a parent node, of which all nodes in the sibling subgraph are
+//! children, as well as a set of edges forming the subgraph boundary.
+//! The boundary must be fully contained within the sibling graph of the parent.
+//!
+//! Sibling subgraphs complement [`super::HierarchyView`]s in the sense that the
+//! latter provide views for subgraphs defined by hierarchical relationships,
+//! while the former provide views for subgraphs within a single level of the
+//! hierarchy.
+//!
+
+use itertools::{Either, Itertools};
+use portgraph::{algorithms::ConvexChecker, view::Subgraph, Direction, PortView};
+use thiserror::Error;
+
+use crate::{
+    ops::{handle::DfgID, OpTag, OpTrait},
+    types::{AbstractSignature, EdgeKind},
+    Hugr, Node, Port, SimpleReplacement,
+};
+
+use super::{sealed::HugrInternals, HugrView};
+
+/// A non-empty convex subgraph of a HUGR sibling graph.
+///
+/// A HUGR region in which all nodes share the same parent. Unlike
+/// [`super::SiblingGraph`],  not all nodes of the sibling graph must be
+/// included. A convex subgraph is always an induced subgraph, i.e. it is defined
+/// by a set of nodes and all edges between them.
+
+/// The incoming boundary (resp. outgoing boundary) is given by the input (resp.
+/// output) ports of the subgraph that are linked to nodes outside of the subgraph.
+/// The signature of the subgraph is then given by the types of the incoming
+/// and outgoing boundary edges. Given a replacement with the same signature,
+/// a [`SimpleReplacement`] can be constructed to rewrite the subgraph with the
+/// replacement.
+///
+/// The ordering of the nodes in the subgraph is irrelevant to define the convex
+/// subgraph, but it determines the ordering of the boundary signature.
+#[derive(Clone, Debug)]
+pub struct SiblingSubgraph<'g, Base> {
+    /// The underlying Hugr.
+    base: &'g Base,
+    /// The nodes of the induced subgraph.
+    nodes: Vec<Node>,
+}
+
+impl<'g, Base: HugrInternals> SiblingSubgraph<'g, Base> {
+    /// A sibling subgraph from a HUGR.
+    ///
+    /// The subgraph is given by the sibling graph of the root. If you wish to
+    /// create a subgraph from another root, wrap the argument `region` in a
+    /// [`super::SiblingGraph`].
+    ///
+    /// This will return an [`InvalidSubgraph::EmptySubgraph`] error if the
+    /// subgraph is empty.
+    pub fn from_sibling_graph(sibling_graph: &'g Base) -> Result<Self, InvalidSubgraph>
+    where
+        Base: HugrView,
+    {
+        let root = sibling_graph.root();
+        let nodes = sibling_graph.children(root).collect_vec();
```

Comment:
Ah, I had missed that `from_dataflow_graph` below skipped the input/output nodes. I'm gonna question whether this should really be a public method though, or something that should live inside `mod tests`. You think whole-subgraph is useful?

Reply:
You're right, will move it to tests.